File: Grids/Cells.m
%% Plotting the grids (Quadtree from Basilisk C dump files)
% Author: Vatsal Sanjay
% vatsalsanjay@gmail.com
% Physics of Fluids
clc
clear
close
%% Output Folder
folder = 'Grids'; % output folder
opFolder = fullfile(cd, folder);
if ~exist(opFolder, 'dir')
mkdir(opFolder);
end
%%
tic
nGFS = 99;
dt = 0.01;
xmin = 0.0; xmin1 = 0.0;
xmax = 1.0; xmax1 = 0.12;
ymin = 0.0; ymin1 = 0.0;
ymax = 1.0; ymax1 = 0.12;
LEVEL = 9;
Xg = {[]};
Xf = {[]};
for ti = 1:1:nGFS
    tic
    t = ti*dt;
    filename = sprintf('intermediate/snapshot-%5.4f',t);
    
    % Grids
    ll=evalc(sprintf('!./getCells %s', filename));
    bolo=textscan(ll,'%f %f\n');
    Xg{ti}.x = bolo{1}; Xg{ti}.y = bolo{2};
    Xg{ti}.x = reshape(Xg{ti}.x, [5, int16(length(Xg{ti}.x)/5)]);
    Xg{ti}.y = reshape(Xg{ti}.y, [5, int16(length(Xg{ti}.y)/5)]);
    % Facets
    ll=evalc(sprintf('!./getFacet %s', filename));
    bolo=textscan(ll,'%f %f\n');
    Xf{ti}.x = bolo{1}; Xf{ti}.y = bolo{2};
    Xf{ti}.x = reshape(Xf{ti}.x, [2, int16(length(Xf{ti}.x)/2)]);
    Xf{ti}.y = reshape(Xf{ti}.y, [2, int16(length(Xf{ti}.y)/2)]);
    
    % plotting
    figure1 = figure('visible','off','WindowState','fullscreen','Color',[1 1 1]);
    axes1 = axes('Parent',figure1);
    hold(axes1,'on');
    name = [folder '/' sprintf('grid%4.4d.png',ti)];
    plot(Xg{ti}.x, Xg{ti}.y,'-','color',[0.5 0.5 0.5],'MarkerSize',30,'LineWidth',1);
    plot(Xf{ti}.x, Xf{ti}.y,'-','color',[0.0 0.0 0.0],'MarkerSize',30,'LineWidth',3);
%     box(axes1,'on');
%     set(axes1,'FontName','times new roman','FontSize',30,'FontWeight','bold',...
%         'LineWidth',3);
%     xlabel('\boldmath{$X$}','LineWidth',2,'FontWeight','bold','FontSize',50,...
%                 'FontName','times new roman',...
%                 'Interpreter','latex');
%     ylabel('\boldmath{$Y$}','LineWidth',2,'FontWeight','bold','FontSize',50,...
%         'FontName','times new roman',...
%         'Interpreter','latex'); 
    axis square
%     xlim([xmin xmax])
%     ylim([ymin ymax])
    xlim([xmin1 xmax1]);
    ylim([ymin1 ymax1]);
    axis off;
    set(figure1,'pos',[1 1 480 480]);
    print(name,'-dpng','-r300')
    close all;
    fprintf('%d of %d\n',ti, nGFS);
    toc
end
%% Saving the data
save('grids','Xg','Xf','-v7.3');
toc

File: Grids/Cells.py
# Author: Vatsal Sanjay
# vatsalsanjay@gmail.com
# Physics of Fluids

# For some reason, python plotting (because of for loop) is really slow
# Use MATLAB as then no need to take care of for loop

import numpy as np
import os
import subprocess as sp
import matplotlib.pyplot as plt
from matplotlib import rc
import matplotlib
import time

matplotlib.rcParams['text.usetex'] = True
matplotlib.rcParams['text.latex.preamble'] = [r'\boldmath']
matplotlib.rcParams['text.latex.unicode'] = True

def gettingCells(filename):
    print('Getting Cells values')
    exe = ["./getCells", filename]
    p = sp.Popen(exe, stdout=sp.PIPE, stderr=sp.PIPE)
    stdout, stderr = p.communicate()
    temp1 = stderr.decode("utf-8")
    temp2 = temp1.split("\n")
    Xtemp = []
    Ytemp = []
    for n1 in range(len(temp2)):
        temp3 = temp2[n1].split(" ")
        if temp3 == ['']:
            pass
        else:
            Xtemp.append(float(temp3[0]))
            Ytemp.append(float(temp3[1]))
    X = np.asarray(Xtemp)
    Y = np.asarray(Ytemp)
    print('Got Cells values')
    return X, Y


def gettingFacets(filename):
    print('Getting facets values')
    exe = ["./getFacet", filename]
    p = sp.Popen(exe, stdout=sp.PIPE, stderr=sp.PIPE)
    stdout, stderr = p.communicate()
    temp1 = stderr.decode("utf-8")
    temp2 = temp1.split("\n")
    Xtemp = []
    Ytemp = []
    for n1 in range(len(temp2)):
        temp3 = temp2[n1].split(" ")
        if temp3 == ['']:
            pass
        else:
            Xtemp.append(float(temp3[0]))
            Ytemp.append(float(temp3[1]))
    X = np.asarray(Xtemp)
    Y = np.asarray(Ytemp)
    print('Got facets values')
    return X, Y

# ----------------------------------------------------------------------------------------------------------------------

nGFS = 2
dt = 0.01
xmin = 0.0
xmax = 1.0
ymin = 0.0
ymax = 1.0
LEVEL = 9

folder = 'Grid'  # output folder
if not os.path.isdir(folder):
    os.makedirs(folder)
for ti in range(1,nGFS):
    t = ti*dt
    place = "intermediate/snapshot-%5.4f" % t
    name = "grid%6.6d.eps" % ti

    Xf, Yf = gettingFacets(place)
    Xg, Yg = gettingCells(place)

    tstart = time.time()
    ## Part to plot
    plt.close()
    fig, ax = plt.subplots()
    fig.set_size_inches(9.40, 9.40)
    rc('axes', linewidth=2)
    plt.xticks(fontsize=30)
    plt.yticks(fontsize=30)

    Xg.resize((int(len(Xg)/5),5))
    Yg.resize((int(len(Yg)/5),5))
    for i in range(0,len(Xg)):
        ax.plot(Xg[i,1:5],Yg[i,1:5],'-',color='grey',linewidth=1)
    ## Drawing Facets
    for i in range(0,len(Xf),2):
        ax.plot([Xf[i], Xf[i+1]],[Yf[i], Yf[i+1]],'-',color='black',linewidth=3)
    # ax.plot(Xg, Yg,'.',color='black',linewidth=3)

    ax.set_xlabel(r'$X$', fontsize=30)
    ax.set_ylabel(r'$Y$', fontsize=30)
    ax.set_aspect('equal')
    ax.set_ylim(xmin, xmax)
    ax.set_xlim(ymin, ymax)
    ax.set_title(r'\textbf{t = %2.1f}' % t, fontsize=30)
    print('Time:', (time.time()-tstart))
    print("Done")
    # plt.show()
    plt.savefig(name)
    plt.close()
